### PR TITLE
[5.0][Feature] Create & update in modal, with AJAX

### PR DESCRIPTION
**Pull request by Gaspertrix**
Just include one line inside you _EntityCrudController::setUp()_ to enable ajax feature:
`$this->crud->enableAjaxCrud();`

You must re publish public assets:
`php artisan vendor:publish --provider="Backpack\CRUD\CrudServiceProvider" --tag="public" #publish CRUD assets`

Create:
snip_20171208222725

Edit:
snip_20171208222807

With inline ajax validation too:
snip_20171208222747

### CODE REVIEW

**Comment by Gaspertrix** [comment]
@lloy0076 It should work with all fields. JS and CSS files are included with response.

I have tested with most common types.

File: src/PanelTraits/AjaxCrud.php [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    |--------------------------------------------------------------------------
+    */
+
+    /**
+     * @var bool
+     */
+    public $ajax_crud = false;
+
+    /**
+     * Check if ajax is enabled for crud.
```
Comment (lloy0076) [inline src/PanelTraits/AjaxCrud.php]: AJAX (consistency).

File: src/app/Http/Controllers/CrudController.php [hunk 2]
```diff
@@ -118,12 +119,15 @@ public function storeCrud(StoreRequest $request = null)
...
+        $message = trans('backpack::crud.insert_success');
+        \Alert::success($message)->flash();
 
         // save the redirect choice for next time
         $this->setSaveAction();
 
-        return $this->performSaveAction($item->getKey());
+        $redirectUrl = $this->performSaveAction($item->getKey());
+
+        return \Request::instance()->ajax() ? response()->json(['redirect_url' => $redirectUrl->getTargetUrl(), 'message' => $message]) : \Redirect::to($redirectUrl);
```
Comment (lloy0076) [inline src/app/Http/Controllers/CrudController.php]: Isn't `$request` already here?

Also can we remove the ternary here and just use `if {}` type clauses.

File: src/app/Http/Controllers/CrudController.php [hunk 3]
```diff
@@ -179,12 +184,15 @@ public function updateCrud(UpdateRequest $request = null)
         $this->data['entry'] = $this->crud->entry = $item;
 
         // show a success message
+        $message = trans('backpack::crud.update_success');
         \Alert::success(trans('backpack::crud.update_success'))->flash();
```
Comment (lloy0076) [inline src/app/Http/Controllers/CrudController.php]: Might as well change this to use `$message`.

File: src/app/Http/Controllers/CrudController.php [hunk 4]
```diff
@@ -179,12 +184,15 @@ public function updateCrud(UpdateRequest $request = null)
...
 
         // save the redirect choice for next time
         $this->setSaveAction();
 
-        return $this->performSaveAction($item->getKey());
+        $redirectUrl = $this->performSaveAction($item->getKey());
+
+        return \Request::instance()->ajax() ? response()->json(['redirect_url' => $redirectUrl->getTargetUrl(), 'message' => $message]) : \Redirect::to($redirectUrl);
```
Comment (lloy0076) [inline src/app/Http/Controllers/CrudController.php]: Remove the ternary - what about the `$request` that already exists.

File: src/public/crud/js/ajax.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+                        });
+                        break;
+                    case 'save_and_back':
+                    default:
+                        $("#modal-ajax-crud").modal("hide");
+                        break;
+                }
+            }
+        },
+        error: function(xhr, statusText, error) {
```
Comment (lloy0076) [inline src/public/crud/js/ajax.js]: This doesn't seem to fire?

File: src/resources/views/edit.blade.php [hunk 6]
```diff
@@ -16,16 +17,17 @@
 
 @section('content')
 <div class="row">
-	<div class="col-md-8 col-md-offset-2">
+	<div @if (!$request->ajax())class="col-md-8 col-md-offset-2"@endif>
 		<!-- Default box -->
-		@if ($crud->hasAccess('list'))
+		@if ($crud->hasAccess('list') AND !$request->ajax())
```
Comment (lloy0076) [inline src/resources/views/edit.blade.php]: What's that `AND`?

File: src/resources/views/form_content.blade.php [hunk 7]
```diff
@@ -28,9 +28,10 @@
     <!-- CRUD FORM CONTENT - crud_fields_scripts stack -->
     @stack('crud_fields_scripts')
 
-    <script>
-    jQuery('document').ready(function($){
-
+    <script type="text/javascript">
```
Comment (lloy0076) [inline src/resources/views/form_content.blade.php]: Ok, this is horrid.

For sanity sake, can we do a:

```
@if IT_IS_AJAX
<script>...</script>
@else
<script>...</script>
@endif
```

**Comment by Gaspertrix** [comment]
@lloy0076 I forgot a very big step, you must re publish public assets so ajax.js is available

File: src/public/crud/js/ajax.js (hunk 5 above)
Comment (Gaspertrix) [reply src/public/crud/js/ajax.js]: https://github.com/Laravel-Backpack/CRUD/pull/1149#issuecomment-351241020

File: src/resources/views/edit.blade.php (hunk 6 above)
Comment (Gaspertrix) [reply src/resources/views/edit.blade.php]: It is a bit redundant to display a link to the list inside the form. So it is hidden inside the ajax form.

Comment (lloy0076) [reply src/resources/views/edit.blade.php]: I meant why `AND` vs `&&`?

**Comment by tabacitu** [comment]
@Gaspertrix , this looks like some excellent work - and it's definitely the direction Backpack is going in. But I'm afraid it's not going to make it into 3.4 - I'm launching it without it this week.

I'll take another look at this in 12 days, and see how we can include it. I expect there's a lot to talk about and modify, since it's such a big change.

Thanks a lot for this. I'll be back.

/remind me to take another look at this in 12 days

**Comment by niladam** [comment]
hi @tabacitu -- i think this needs to be looked at ? -- just a friendly reminder 👍

**Comment by niladam** [comment]
hey @tabacitu -- it's been 15 days since i nudged you and **61** days since your request to be reminded in 12 days. Has this been abandoned ? What are your thoughts on this ?

**Comment by tabacitu** [comment]
Haha - thank you for the reminders @niladam . On a second look, this looks like it's out of scope for ```3.5``` - will have another look at it and definitely have it in ```4.0```.

Sorry this did not get merged yet @Gaspertrix - please don't take it personally. I myself have dozens of PRs that have not been merged yet, and dozens more that got closed without being merged. It's just that we can't always find the opportunity to test, polish, merge and document all features - we have a priority list and can only tackle top priorities for each version. Time is limited, I'm afraid.

**Comment by ziming** [comment]
@tabacitu onkbear recently released this package for backpack 4

https://github.com/onkbear/backpack-nested-crud

Maybe worth a look. I have not tried it myself yet.

**Comment by pxpm** [comment]
@ziming just pushed PR #2280 :) omg everybody wants this so bad! With so many options i think we could achieve the best of it! Let's keep working on this! 

Best,
Pedro

**Comment by tabacitu** [comment]
Ugh... how did we not close this before? Thank you for the PR @Gaspertrix , it was instrumental for @pxpm in creating the feature. We have it since 4.1 I think 💪

**Comment by tabacitu** [comment]
Wait, no. It's not the same thing, sorry 🙏

**Comment by tabacitu** [comment]
@Gaspertrix thanks for submitting this, but I'm afraid this is a bit too much for Backpack, it would add too much complexity.

I do agree with the feature, though. I want it too. So we'll create the new Livewire components in a way that that can be included in a page/modal/drawer just as easily. 

That, I think, would be a better long-term solution, creating those new components with reusability in mind, rather than bending & twisting our current components into doing something they weren't designed to do.

Again, thanks a lot for taking the time to submit this. We super appreciate it, and apologise for taking so long to reach this conclusion. We've been super-swamped in the past, but now that out team has gotten a little bigger, we're moving much _much_ faster.

Cheers!

**Pull request closed without merging**